tests/upgrades/test_upgrades.py

```diff
@@ -0,0 +1,325 @@
+            admin.contract_address
+        ]
+    )
+
+    # upgrade
+    tx_exec_info = await signer.send_transaction(
+        admin, proxy.contract_address, 'upgrade', [
+            v2.contract_address
+        ]
+    )
+
+    # check event
+    assert_event_emitted(
+        tx_exec_info,
+        from_address=proxy.contract_address,
+        name='Upgraded',
+        data=[
+            v2.contract_address
+        ]
+    )
+
+
+@pytest.mark.asyncio
+async def test_upgrade_from_non_admin(proxy_factory):
+    admin, non_admin, _, v2, _, proxy = proxy_factory
+
+    # initialize implementation
+    await signer.send_transaction(
+        admin, proxy.contract_address, 'initializer', [
+            admin.contract_address
+        ]
+    )
+
+    # upgrade should revert
+    await assert_revert(
+        signer.send_transaction(
+            non_admin, proxy.contract_address, 'upgrade', [
+                v2.contract_address
+            ]
+        ),
+        reverted_with="Proxy: caller is not admin"
+    )
+
+
+@pytest.mark.asyncio
+# Using `after_upgrade` fixture henceforth
+async def test_implementation_v2(after_upgrade):
+    admin, _, _, v2, _, proxy = after_upgrade
+
+    # check implementation address
+    execution_info = await signer.send_transaction(
+        admin, proxy.contract_address, 'get_implementation', []
+    )
+    assert execution_info.result.response == [v2.contract_address]
+
+    # check admin
+    execution_info = await signer.send_transaction(
+        admin, proxy.contract_address, 'get_admin', []
+    )
+    assert execution_info.result.response == [admin.contract_address]
+
+    # check value
+    execution_info = await signer.send_transaction(
+        admin, proxy.contract_address, 'get_value_1', []
+    )
+    assert execution_info.result.response == [VALUE_1, ]
+
+
+@pytest.mark.asyncio
+async def test_set_admin(after_upgrade):
+    admin, new_admin, _, _, _, proxy = after_upgrade
+
+    # change admin
+    await signer.send_transaction(
+        admin, proxy.contract_address, 'set_admin', [
+            new_admin.contract_address
+        ]
+    )
+
+    # check admin
+    execution_info = await signer.send_transaction(
+        admin, proxy.contract_address, 'get_admin', []
+    )
+    assert execution_info.result.response == [new_admin.contract_address]
+
+
+@pytest.mark.asyncio
+async def test_set_admin_from_non_admin(after_upgrade):
+    _, non_admin, _, _, _, proxy = after_upgrade
+
+    # change admin should revert
+    await assert_revert(
+        signer.send_transaction(
+            non_admin, proxy.contract_address, 'set_admin', [
+                non_admin.contract_address
+            ]
+        )
+    )
+
+
+@pytest.mark.asyncio
+async def test_implementation_v3(after_upgrade):
+    admin, _, _, _, v3, proxy = after_upgrade
+    # The v3 implementation sets a new variable in storage in v2
+    # and this tests that both v1 and v2's storage carries over to v3
```

Comment:
but isn't v2 and v3 the same? I'm not sure I get the purpose

Reply:
Indeed, they are. It might be redundant, but I wanted to check that the variables set in  `v1` and `v2` carry over to `v3` correctly. Do you think it's worth making a stand alone `v3`? Or maybe scrapping it?

Reply:
Yeah, that's fair. I suppose it's more of a sanity check for me initially. I'll take it out